File: 110-balanced-binary-tree/110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # returns -1 if the tree is not balanced
    def getBalancedHeight(self, node: Optional[TreeNode]) -> int:
        if node == None:
            return 0
        
        left = self.getBalancedHeight(node.left)
        right = self.getBalancedHeight(node.right)
        # print("{0}, {1} -> {2}".format(left, right, abs(right - left)))
        
        if left == -1 or right == -1:
            return -1
        
        
        if (abs(right - left) > 1 ):
            return -1
        
        return max(left, right) + 1
        
        
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        height = self.getBalancedHeight(root)
        if height == -1:
            return False
        
        return True

File: 110-balanced-binary-tree/110-balanced-binary-tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }

use std::rc::Rc;
use std::cell::RefCell;
use std::cmp::max;

impl TreeNode{
    pub fn dfs_height(&self)-> i32{
        let mut lh = 0;
        let mut rh = 0;
        
        if let Some(ref left) = self.left{
            lh = left.borrow().dfs_height();
            if lh == -1 {
                return -1;
            }
        }
        
        if let Some(ref right) = self.right{
            rh = right.borrow().dfs_height();
            if rh == -1 {
                return -1;
            }
        }
        
        if (lh - rh).abs() > 1 {
            -1
        }else{
            max(rh, lh)  + 1
        }
    }
}

impl Solution {
    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        if let Some(ref node) = root {
            let h = node.borrow().dfs_height();
            h != -1
        }else{
            true
        }
    }
}